[llvm-lit] Adding -e option to lit's built-in cat command

connieyzhu opened the pull request:
This patch extends support for using cat -e in lit tests that use the internal
shell. The logic for enabling options is changed to accommodate for the
possibility of having several options. This patch is dependent on https://github.com/llvm/llvm-project/pull/101530. 

Fixes https://github.com/llvm/llvm-project/issues/102377

connieyzhu committed:
[llvm-lit] Adding -e option to lit's built-in cat command

This patch extends support for using cat -e in lit tests that use the internal
shell. The logic for enabling options is changed to accommodate for the
possibility of having several options.

connieyzhu committed:
[llvm-lit] Restructure the tracking of enabled cat options

This patch changes the storing of enabled options from a list of bools to a
dataclass for better efficiency when accessing and checking the options.

connieyzhu committed:
[llvm-lit] Fixed darker formatting issue

ilovepi commented on llvm/utils/lit/lit/builtin_commands/cat.py:
```diff
@@ -69,8 +87,8 @@ def main(argv):
 
-            if show_nonprinting:
-                contents = convertToCaretAndMNotation(contents)
+            if True in fields(enabled_options):
+                contents = convertTextNotation(contents, enabled_options)
```
What happens if I add a field to `Options` that isn't either `show_nonprinting` or `show_ends`?

connieyzhu replied:
If both `show_nonprinting` and `show_ends` are False and this added field was True, `convertTextNotation` has an `else` statement on lines 43/44 that should account for this case, and not change the text (assuming that there is no existing implementation for this new option). But this is just a precaution, because I don't think this case should technically ever occur though - any option that exists in `Options` should have an implementation in `convertTextNotation`. Does this address your concern or am I misunderstanding?

ilovepi replied:
But that means well run the conversion code when we don't need to right? It's preferable in my opinion to avoid that work if we don't have to, specially since that routine a character by character copy.

I'm also wondering if we should have an early exit from the routine if the neither option is set. WDYT?

ilovepi commented on llvm/utils/lit/lit/builtin_commands/cat.py:
```diff
@@ -1,29 +1,45 @@
-            newdata.write(chr(intval + 64))
-        elif intval == 127:
-            newdata.write("^?")
+        if options.show_ends:
+            if intval == 10:
```
nit: you can combine these and drop a level of nesting. Could also be `if/else if/else` but that isn't too important since there's a continue.

connieyzhu replied:
done

ilovepi reviewed (commented):
Mostly OK, with some minor comments, but don't the tests from #101530 need to be updated in this patch?

connieyzhu commented:
> Mostly OK, with some minor comments, but don't the tests from #101530 need to be updated in this patch?

I wouldn't be able to update the tests in this patch until https://github.com/llvm/llvm-project/pull/101530 has landed right, since I would need to rebase first?

connieyzhu committed:
[llvm-lit][NFC] Added comments for dataclass + simplified show_ends
logic

connieyzhu commented on llvm/utils/lit/lit/builtin_commands/cat.py:
```diff
@@ -1,29 +1,45 @@
+@dataclass
+class Options:
+    show_ends: bool
+    show_nonprinting: bool
```
done

ilovepi reviewed (commented):
Mostly OK, with some minor comments, but don't the tests from #101530 need to be updated in this patch?

connieyzhu committed:
[llvm-lit] Added variable to track when cat should change text notation

This patch adds a new bool variable convert_text that is turned on if
show_nonprinting, show_ends, or any option added in the future that is
meant to modify text is True. This fixes the faulty logic that assumes
that any option that is True is one that will modify text. Now,
convertTextNotation() will only be called for options that need it.

jh7370 reviewed (commented):
Change looks basically fine to me (once tests are updated). One question I do have is how many tests actually use this functionality and do they really need to?

connieyzhu commented:
> One question I do have is how many tests actually use this functionality and do they really need to?

@jh7370 I believe only one test currently uses cat -e. I'm not quite sure what its purpose is, but it's in [TestUseColor.test](https://github.com/llvm/llvm-project/blob/main/lldb/test/Shell/Driver/TestUseColor.test#L4) in `lldb`.

connieyzhu committed:
[llvm-lit] Added extra check when calling convertTextNotation

This patch adds an assert to guarantee that when calling
convertTextNotation(), the show_ends and show_nonprinting options are
turned on. Some logic is also simplified when converting characters.

connieyzhu committed:
[llvm-lit] Added -E option to built-in cat

This patch allows for lit's built-in cat to use the -E option. The
functionality was already implemented in previous patches, but this
patch allows for the parsing and usage of -E.

jh7370 commented:
> > One question I do have is how many tests actually use this functionality and do they really need to?
> 
> @jh7370 I believe only one test currently uses cat -e. I'm not quite sure what its purpose is, but it's in [TestUseColor.test](https://github.com/llvm/llvm-project/blob/main/lldb/test/Shell/Driver/TestUseColor.test#L4) in `lldb`.

Thanks, I don't see the need for the `$` line ending behaviour in that test. It's possible the non-printing characters behaviour is needed though. @JDevlieghere, since you last touched that test, can you confirm the purpose of the -e usage is to print non-printing characters?

We should avoid adding functionality that isn't needed, so for example, -v might be more appropriate in that test, which would then allow us to avoid adding -e functionality.

petrhosek commented:
If we decide to go ahead with this change, I'd consider replacing the use of `getopt` with `argparse` which is more idiomatic in modern Python and should simplify the implementation.

connieyzhu commented:
Decided to go with the solution in https://github.com/llvm/llvm-project/pull/104878, so I will be closing this PR.
